### ISSUE TITLE
Model field to filter detection inconsistent with other libs

### ISSUE THREAD

**Issue by stevelacey** [philipn/django-rest-framework-filters#102]
**tl;dr** I expect these two classes to yield the same result and they don't

```
class InspectionFilter(filters.FilterSet):
    """filtering by inspection_date will be wonderful"""
    inspection_date = filters.DateTimeFilter(name='inspection_date')

    class Meta:
        model = Inspection
```

```
class InspectionFilter(filters.FilterSet):
    """filtering by inspection_date will be a sad ISO-orientated experience"""
    class Meta:
        model = Inspection
        fields = ('inspection_date',)
```

I don't know if same applies to other fields, still playing – I suggest a good improvement would be to introspect the model classes like how admin/forms/drf serializers do – and apply sensible filter defaults.

**Comment by stevelacey** [comment]
Scratch that I am not sure you even do anything with dates despite the note in the README?

I want non ISO filters – which I guess must exist in drf-filters now since I don't see you patching it anywhere(/anymore?) and it works with the former – it'd also be nice to have AllLookupsFilter functionality on dates e.g. `inspection_date__gte=2015-01-01`… which I assumed existed, but don't appear to be able to find the right syntax for.

**Comment by stevelacey** [comment]
I realised I can manually specify each additional lookup myself using the `DateTimeFilter`. I wonder if there is value in providing more magic here, I as a consumer of this library expected `AllLookupsFilter` to magically yield a bunch of extra filter possibilities based on the data type I apply it to – I realise now that is not the case; and the library is a little less opinionated than that, which is arguably a good thing.

I think it would be neat to offer some more opinionated options, if there was a fitler to dump on say date types (automatically detected) – that allowed for things like `inspection_date__before/lte`, `inspection_date__after/gte` without any additional effort required, that'd be pretty powerful stuff.

**Comment by rpkilby** [comment]
In your first example, you manually declare a localized `DateTimeFilter`. When using the `Meta.fields` option, the FilterSet introspects the model field and chooses a suitable filter class. `IsoDateTimeFilter`s are preferred for datetime model fields, which is why you have a discrepancy.

> I suggest a good improvement would be to introspect the model classes like how admin/forms/drf serializers do

Not sure what makes you think that this isn't what happens. django-filter/drf-filters performs the same field [introspection](https://github.com/carltongibson/django-filter/blob/0.14.0/django_filters/filterset.py#L196-L283) as DRF's [serializers](https://github.com/tomchristie/django-rest-framework/blob/3.4.5/rest_framework/serializers.py#L792-L817).

> and apply sensible filter defaults.

ISO date/datetime values are sensible defaults. The standard exists for a reason.
- ISO datetimes are not ambiguous. (is 2/1/2016 February first or January second? This is a problem with [localized date formats](https://en.wikipedia.org/wiki/Date_format_by_country).)
- ISO datetimes have a standard timezone handling. 

> I want non ISO filters 

Why? If this is a UI issue, I suggest that your app/client transform a localized datetime into an ISO datetime and pass that to the API. Keep in mind that django-filter is intended to produce a form suitable for a user interface. In this case, it makes sense to use a localized datetime field. drf-filters is intended to be used in an API context, and should not be user-facing. The consumer here is a machine, not a human.

> which I guess must exist in drf-filters now since I don't see you patching it anywhere(/anymore?) and it works with the former

No patching, but the filterset is [overridden](https://github.com/philipn/django-rest-framework-filters/blob/master/rest_framework_filters/filterset.py#L126-L138) to use `IsoDateTimeFilter`s as the default for datetime fields.

> it'd also be nice to have AllLookupsFilter functionality on dates e.g. `inspection_date__gte=2015-01-01`

Dates are not datetimes, which is why the filtering by date fails. Django's filter syntax has a concept of transforms. To filter by datetimes by date values, you can use `inspection_date__date__lte=2015-01-01`. The syntax looks weird, but that's in part a result of the datetime field being named as a date (Might have chosen `inspected_at` instead, resulting in `inspected_at__date__lte`). If the query param is unacceptable, you can always manually declare filters. ex:

```py
class InspectionFilter(FilterSet):
    inspection_date = filters.DateFilter(name='inspection_date', lookup_expr='date')
    inspection_date__before = filters.DateFilter(name='inspection_date', lookup_expr='date__lte')

# /inspections?inspection_date=2015-01-01 translates to
Inspection.objects.filter(inspection_date__date=date(2015, 1, 1))

# /inspections?inspection_date__before=2015-01-01 translates to
Inspection.objects.filter(inspection_date__date__lte=date(2015, 1, 1))
```

> I as a consumer of this library expected AllLookupsFilter to magically yield a bunch of extra filter possibilities based on the data type I apply it to – I realise now that is not the case; and the library is a little less opinionated than that, which is arguably a good thing.

As explained above, it does. `AllLookupsFilter` does basically two things. Given a model field, it will [introspect](https://github.com/philipn/django-rest-framework-filters/blob/v0.8.0/rest_framework_filters/utils.py#L9-L24) all registered lookups for that field's type. For each lookup, it will automatically [generate](https://github.com/philipn/django-rest-framework-filters/blob/v0.8.0/rest_framework_filters/filterset.py#L59-L63) a new filter for the field type and lookup. This uses the same mechanisms as django-filter's `Meta.fields` option.

> I think it would be neat to offer some more opinionated options, if there was a fitler to dump on say date types (automatically detected) – that allowed for things like inspection_date__before/lte

I've thought about it, but replacing `lte` with `before` is dependent on type, which we can only guess so well. e.g., it does not make sense to use `before` with number types. In the end, consistency between types (all fields use `lte` instead of a `lte`/`before` mishmash) and consistency w/ Django's API made the most sense. After all, drf-filters is intended to be interfaced programmatically.

**Comment by rpkilby** [comment]
Forgot a few things:

> Model field to filter detection inconsistent with other libs

What libraries and what's inconsistent? For the ISO datetime issue, DRF serializers use the [ISO format] (https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/settings.py#L103-L104) too.

----

In general, not sure if there's anything to do here? There seems to be two main problems from above:

**drf-filters should use localized datetimes instead of the ISO 8601 format**
The ISO 8601 format is preferred for most cases, so this will not change.
- The correct thing to do is have your app transform localized, displayable datetimes into ISO datetimes. 
- You can extend the filter field's input formats, as explained in the [docs](https://django-filter.readthedocs.io/en/latest/ref/fields.html#isodatetimefield).
- If you really need/want to use just localized datetimes, I recommend looking at [`filter_overrides`](https://django-filter.readthedocs.io/en/latest/usage.html#overriding-default-filters).

**Unable to filter datetime fields by dates**
As mentioned above, in order to filter datetimes by dates, you have to use the `date` transform. Alternatively, you can add date formats to `IsoDateTimeField.input_formats`.

**Comment by stevelacey** [comment]
Thanks for the thorough write-up, you misinterpreted a couple of my points, but answered all my questions, just to revise:

- I wasn't suggesting switching `lte/gte` to `before/after` – based on my crude research I didn't think these existed via `AllLookupsFilter` – I now understand this is because I was using the wrong filter class
- I agree that ISO datetimes are awesome and should be everywhere – I absolutely want to support these foremost
- I am torn, I am of the opinion that whilst APIs should always output ISO datetimes, filtering should be easy and when they don't care shouldn't have to write out an ISO datetime to use them – in this case I would expect the datetime to simply be assumed to be UTC; but I can understand why others may disagree, but how would you even interpret a localized otherwise – you'd have to pass a `timezone` query param along for the ride… which actually I have done before (for a frontend) and worked pretty well
- RE: patching, ah, gotcha noted
- RE `_date` ah, yes, I normally go for `_at`, but my client (a developer) designed the schema and chose `_date`… I couldn't remember why I prefer `_at` – thanks for an excuse to change it 😉 

--

In summary:

- Yes, no changes required, I am a thicko, a little more docs on datetime filters and gte/lte support might be a good addition for the wiki, otherwise, cool 👌
- I am yet to try any of this but based on what you've said I think I want to switch back to inferring the type from models by using `Meta.field` and using an ISODateTimeFilter, and either supply ISO datetimes when filtering – or mess with `IsoDateTimeField.input_formats`… and rename my columns to avoid `?inspection_date__date__bleugh` 😄  thanks for your time
